crates/rspack_binding_values/src/dependency.rs

```diff
@@ -1,108 +1,193 @@
+use std::{cell::RefCell, ptr::NonNull};
+
+use napi::bindgen_prelude::ToNapiValue;
 use napi_derive::napi;
-use rspack_core::{BoxDependency, DependencyId};
+use rspack_core::{Compilation, CompilationId, Dependency, DependencyId};
+use rspack_napi::OneShotRef;
+use rustc_hash::FxHashMap as HashMap;
 
 // JsDependency allows JS-side access to a Dependency instance that has already
 // been processed and stored in the Compilation.
 #[napi]
-pub struct JsDependency(&'static BoxDependency);
-
-impl JsDependency {
-  pub(crate) fn new(dependency: &BoxDependency) -> Self {
-    // SAFETY:
-    // The lifetime of the &mut BoxDependency reference is extended to 'static.
-    // Accessing fields and methods on the Rust object from the JS side after the Rust object's
-    // lifetime has ended is undefined behavior, which we currently disregard.
-    let dependency =
-      unsafe { std::mem::transmute::<&BoxDependency, &'static BoxDependency>(dependency) };
-    Self(dependency)
-  }
-
-  pub(crate) fn id(&self) -> &DependencyId {
-    self.0.id()
-  }
+pub struct JsDependency {
+  pub(crate) compilation: Option<NonNull<Compilation>>,
+  pub(crate) dependency_id: DependencyId,
+  pub(crate) dependency: NonNull<dyn Dependency>,
 }
 
-#[napi]
 impl JsDependency {
-  #[napi(getter)]
-  pub fn get_type(&self) -> &str {
-    self.0.dependency_type().as_str()
+  fn attach(&mut self, compilation: NonNull<Compilation>) {
+    if self.compilation.is_none() {
+      self.compilation = Some(compilation);
+    }
   }
 
-  #[napi(getter)]
-  pub fn category(&self) -> &str {
-    self.0.category().as_str()
-  }
+  fn as_ref(&mut self) -> napi::Result<&dyn Dependency> {
+    let dependency = unsafe { self.dependency.as_ref() };
```

Comment:
safety comments ?

Reply:
The JS API is now in a fixed pattern and will use ids to verify pointers, if invalid will return Err. So I think safe comments are unnecessary.